content/2015.09.kinto-encryption-example.rst

```diff
@@ -0,0 +1,207 @@
+How to encrypt data with Kinto.js
+###########################################
+
+:date: 2015-09-16
+:summary: How to encrypt data with Kinto.js.
+:slug: kinto-encryption-example
+:lang: en
+:url: kinto-encryption-example
+
+A guest post this time, by Michiel de Jong from the Firefox OS team! Sorry for
+writing in English. :)
+
+If you want to encrypt data that is uploaded to the cloud by an offline-first app,
+it makes sense to keep the local copy of the data in-the-clear, and encrypt the
+data only just before it is uploaded. That way, you can easily and efficiently display
+and update the local copy of the data, without the repeated cost of encrypting
+and decrypting everything all the time. Likewise, you only need to decrypt incoming data
+once, when it's downloaded, and your app can then display the data as often as
+needed without extra processing costs.
+
+Kinto.js now supports
+`<code>RemoteTransformer</code>s http://kintojs.readthedocs.org/en/latest/api/#transformers`_,
+which are superhandy for this pattern.
+This blogpost shows how to use a <code>RemoteTransformer</code> for encrypting your data just
+before it's sent to the cloud. It uses WebCrypto for the cryptography.
+
+The main idea behind a <code>RemoteTransformer</code> is that you provide two functions,
+"encode" and "decode", and Kinto.js makes sure to call the encode function on
+each record that gets uploaded, and to call the decode function on each record
+that gets downloaded. You can also read about this in the Kinto.js documentation,
+http://kintojs.readthedocs.org/.
+
+NB: I am not a security expert and I wrote this example in an afternoon, so
+if you're going to copy-paste this code, please don't use it for anything
+valuable without doing your own proper security review first! :)
```

Comment:
I thought about that, but there are a few considerations:
* This code is quite similar to how FxSync clients work, except that I left out the hmac signing/verifying. I'm not even sure why hmac on top of aes is important, but I'm sure it's there for a reason, so this fact alone would probably be enough to say that this code is not safe to use in production.
* Security depends on how, and for what, this snippet of code is used; there's no sensible way to define the threat models of a snippet without knowing more about its context.
* The [feedback we got](https://bugzilla.mozilla.org/show_bug.cgi?id=1191776#c13) on the crypto part of the sync app mentioned that key generation requires "careful review to make sure the keys aren't weak". The sync app does not do key generation (yet), this snippet does. I don't feel it's worth people's time to do such a careful review on a snippet that does not even do the hmac-checking part.
* Adding a disclaimer like this is an easy way to publish this example code and for me to still sleep well at night without worrying about people relying on my snippet to be usable for anything else than fun and experimentation. :)
* People can still use this snippet as a basis for "real" projects, and if they do, they will want to add hmac-verification, invent a way to persist `aesKey` (currently a new one is generated on each run) and do their own security review of their full product anyway. And then they can still give feedback on the snippet code if they find any security bugs in the minimal things (generate key, generate IV, encrypt, decrypt) that this snippet does do.